# Notebook 02 — Limpieza y Calidad de Datos
**Proyecto Integrador | Minería de Datos I**

## Objetivo
Aplicar un pipeline de limpieza completo sobre el dataset crudo `streaming_users_dirty` para obtener un dataset libre de duplicados, inconsistencias, outliers y valores faltantes, listo para el análisis exploratorio.

## Etapa
Esta es la etapa 2 de 5. Cada decisión de limpieza está documentada y justificada. El resultado se guarda en `data/processed/streaming_users_clean.csv`.

In [ ]:
# Celda 1
# Paso 1: Cargo el dataset y observo si cargo bien

import pandas as pd

# Cargar el dataset que vamos a limpiar usando rutas relativas
df = pd.read_json("../data/raw/streaming_users_dirty.json")

# Ver que haya cargado bien mostrando las primeras filas
df.head()

,user_id,age,subscription_plan,monthly_watch_time_mins,country,favorite_genre,last_login_date,customer_support_tickets
0,10000,39,Estándar,805.8,Brasil,Crime,2025-03-04,99
1,10001,37,Estándar,1173.4,Colombia,Crime,2019-04-02,2
2,10002,28,Básico,401.0,Colombia,Crime,2018-04-13,0
3,10003,43,Básico,62.4,Uruguay,Thriller,2021-01-31,0
4,10004,51,Básico,477.8,Perú,Thriller,2020-09-30,1


In [ ]:
# Celda 2
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 8160 entries, 0 to 8159
Data columns (total 8 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   user_id                   8160 non-null   int64  
 1   age                       8160 non-null   int64  
 2   subscription_plan         8160 non-null   str    
 3   monthly_watch_time_mins   7967 non-null   float64
 4   country                   8160 non-null   str    
 5   favorite_genre            7920 non-null   str    
 6   last_login_date           7840 non-null   str    
 7   customer_support_tickets  8160 non-null   int64  
dtypes: float64(1), int64(3), str(4)
memory usage: 756.7 KB


In [ ]:
# Celda 3
# ANÁLISIS DE MISSING DATA (MCAR, MAR, MNAR)

print("ANÁLISIS DE MISSING DATA (MCAR / MAR / MNAR)")


# 1. Identificar columnas con nulos
columnas_con_nulos = df.columns[df.isnull().any()].tolist()
print(f"\n📌 Columnas con valores faltantes: {columnas_con_nulos}")

print("\n📌 Porcentaje de nulos por columna:")
print((df.isnull().sum() / len(df) * 100).round(2))


ANÁLISIS DE MISSING DATA (MCAR / MAR / MNAR)

📌 Columnas con valores faltantes: ['monthly_watch_time_mins', 'favorite_genre', 'last_login_date']

📌 Porcentaje de nulos por columna:
user_id                     0.00
age                         0.00
subscription_plan           0.00
monthly_watch_time_mins     2.37
country                     0.00
favorite_genre              2.94
last_login_date             3.92
customer_support_tickets    0.00
dtype: float64


In [ ]:
# Celda 4
# 2. Análisis detallado de MAR
print("\n" + "="*70)
print("ANÁLISIS DE MAR: ¿Los nulos dependen de otras variables?")
print("="*70)

# 2.1 monthly_watch_time_mins vs subscription_plan
print("\n📌 monthly_watch_time_mins vs subscription_plan:")
print("\n▶️ Distribución de subscription_plan en nulos:")
nulos_tiempo = df[df['monthly_watch_time_mins'].isnull()]['subscription_plan'].value_counts(normalize=True).round(3)
print(nulos_tiempo)

print("\n▶️ Distribución de subscription_plan en no nulos:")
no_nulos_tiempo = df[df['monthly_watch_time_mins'].notnull()]['subscription_plan'].value_counts(normalize=True).round(3)
print(no_nulos_tiempo)


ANÁLISIS DE MAR: ¿Los nulos dependen de otras variables?

📌 monthly_watch_time_mins vs subscription_plan:

▶️ Distribución de subscription_plan en nulos:
subscription_plan
Premium      0.762
Estándar     0.161
Premiun      0.026
Básico       0.021
Premium      0.010
STANDARD     0.005
Estándar     0.005
estandar     0.005
premium      0.005
Name: proportion, dtype: float64

▶️ Distribución de subscription_plan en no nulos:
subscription_plan
Básico       0.433
Estándar     0.336
Premium      0.172
basico       0.008
BASICO       0.007
Basic        0.007
básico       0.006
Std          0.006
Estándar     0.006
estandar     0.004
STANDARD     0.004
Premium      0.004
PREMIUM      0.003
premium      0.003
Premiun      0.002
Name: proportion, dtype: float64


In [ ]:
# Celda 5
# 2.2 favorite_genre vs subscription_plan
print("\n📌 favorite_genre vs subscription_plan:")
print("\n▶️ Distribución de subscription_plan en nulos:")
nulos_genero = df[df['favorite_genre'].isnull()]['subscription_plan'].value_counts(normalize=True).round(3)
print(nulos_genero)

print("\n▶️ Distribución de subscription_plan en no nulos:")
no_nulos_genero = df[df['favorite_genre'].notnull()]['subscription_plan'].value_counts(normalize=True).round(3)
print(no_nulos_genero)

# 2.3 last_login_date vs subscription_plan
print("\n📌 last_login_date vs subscription_plan:")
print("\n▶️ Distribución de subscription_plan en nulos:")
nulos_fecha = df[df['last_login_date'].isnull()]['subscription_plan'].value_counts(normalize=True).round(3)
print(nulos_fecha)

print("\n▶️ Distribución de subscription_plan en no nulos:")
no_nulos_fecha = df[df['last_login_date'].notnull()]['subscription_plan'].value_counts(normalize=True).round(3)
print(no_nulos_fecha)


📌 favorite_genre vs subscription_plan:

▶️ Distribución de subscription_plan en nulos:
subscription_plan
Básico      0.450
Estándar    0.333
Premium     0.171
Std         0.012
básico      0.012
basico      0.008
STANDARD    0.004
Premium     0.004
estandar    0.004
Name: proportion, dtype: float64

▶️ Distribución de subscription_plan en no nulos:
subscription_plan
Básico       0.422
Estándar     0.332
Premium      0.187
basico       0.007
BASICO       0.007
Basic        0.007
básico       0.006
Estándar     0.006
Std          0.006
estandar     0.004
STANDARD     0.004
Premium      0.004
PREMIUM      0.003
Premiun      0.003
premium      0.003
Name: proportion, dtype: float64

📌 last_login_date vs subscription_plan:

▶️ Distribución de subscription_plan en nulos:
subscription_plan
Básico      0.453
Estándar    0.319
Premium     0.175
Premium     0.012
STANDARD    0.006
BASICO      0.006
PREMIUM     0.006
básico      0.006
Std         0.006
estandar    0.003
Premiun     0.003
Basic  

In [ ]:
# Celda 6
# 3. Conclusión del análisis
print("\n" + "="*70)
print("CONCLUSIÓN DEL ANÁLISIS DE MISSING DATA")
print("="*70)

print("""
📌 monthly_watch_time_mins:
   - 76.2% de los nulos pertenecen a usuarios Premium.
   - En los no nulos, solo el 19.4% son Premium.
   - Esto indica que el dato faltante DEPENDE del plan de suscripción.
   → Clasificación: MAR (Missing At Random)
   → Estrategia: Imputación por grupo (mediana por plan) o 
                  imputación global con justificación.

📌 favorite_genre:
   - La distribución de nulos vs no nulos es similar entre planes.
   - No se observa dependencia clara con otras variables.
   → Clasificación: MCAR (Missing Completely At Random)
   → Estrategia: Imputación con 'Desconocido' (categoría neutra)

📌 last_login_date:
   - La distribución de nulos vs no nulos es similar entre planes.
   - No se observa dependencia clara.
   → Clasificación: MCAR (Missing Completely At Random)
   → Estrategia: Imputación con fecha mediana global

⚠️ IMPORTANTE: 
   - Al ser MAR, imputar monthly_watch_time_mins con la mediana global 
     podría introducir sesgo. Se recomienda imputar por grupo (mediana por plan).
   - Para este trabajo, se mantendrá la imputación global por simplicidad,
     pero se reconoce esta limitación en las conclusiones finales.
""")


CONCLUSIÓN DEL ANÁLISIS DE MISSING DATA

📌 monthly_watch_time_mins:
   - 76.2% de los nulos pertenecen a usuarios Premium.
   - En los no nulos, solo el 19.4% son Premium.
   - Esto indica que el dato faltante DEPENDE del plan de suscripción.
   → Clasificación: MAR (Missing At Random)
   → Estrategia: Imputación por grupo (mediana por plan) o 
                  imputación global con justificación.

📌 favorite_genre:
   - La distribución de nulos vs no nulos es similar entre planes.
   - No se observa dependencia clara con otras variables.
   → Clasificación: MCAR (Missing Completely At Random)
   → Estrategia: Imputación con 'Desconocido' (categoría neutra)

📌 last_login_date:
   - La distribución de nulos vs no nulos es similar entre planes.
   - No se observa dependencia clara.
   → Clasificación: MCAR (Missing Completely At Random)
   → Estrategia: Imputación con fecha mediana global

⚠️ IMPORTANTE: 
   - Al ser MAR, imputar monthly_watch_time_mins con la mediana global 
     podr

In [ ]:
# Celda 7
# Imputación por grupo (recomendado para MAR)
print("\n📌 Imputación de monthly_watch_time_mins por grupo (plan):")
print("Medianas por plan:")
print(df.groupby('subscription_plan')['monthly_watch_time_mins'].median())

# Imputar usando transform
df['monthly_watch_time_mins'] = df.groupby('subscription_plan')['monthly_watch_time_mins'].transform(
    lambda x: x.fillna(x.median())
)


📌 Imputación de monthly_watch_time_mins por grupo (plan):
Medianas por plan:
subscription_plan
BASICO        445.75
Basic         582.65
Básico        554.15
Estándar      840.65
Estándar      755.40
PREMIUM      1239.45
Premium      1127.25
Premium      1079.60
Premiun      1058.80
STANDARD      799.70
Std           873.90
basico        476.70
básico        596.95
estandar      855.20
premium      1258.70
Name: monthly_watch_time_mins, dtype: float64


### Paso 2: Eliminación de Duplicados Exactos
Se identificaron 126 filas completamente idénticas en todas sus columnas, producto de errores en la exportación original del dataset. Se eliminan usando `keep='first'` para conservar la primera aparición y descartar las repetidas.

In [ ]:
# Celda 9
# Paso 2: Eliminación de duplicados exactos
filas_antes = df.shape[0]

df = df.drop_duplicates(keep='first')

filas_despues = df.shape[0]
print(f"--- RESULTADO PASO 2 ---")
print(f"Filas eliminadas: {filas_antes - filas_despues}")
print(f"Dimensiones actuales: {df.shape[0]} filas y {df.shape[1]} columnas.")

--- RESULTADO PASO 2 ---
Filas eliminadas: 126
Dimensiones actuales: 8034 filas y 8 columnas.


### Paso 3: Resolución de Claves Duplicadas (user_id)
El `user_id` debe ser único por usuario. Se detectaron 34 usuarios con más de un registro con datos inconsistentes entre sí. Se conserva el último registro de cada usuario con `keep='last'`, asumiendo que el registro más reciente en el archivo refleja el estado actual del usuario.

In [ ]:
# celda 11
# Paso 3: Resolución de user_id duplicados - se mantiene el último registro
filas_antes = df.shape[0]
id_duplicados_antes = df[df.duplicated(subset=['user_id'], keep=False)]['user_id'].nunique()

df = df.drop_duplicates(subset=['user_id'], keep='last')

filas_despues = df.shape[0]
print(f"--- RESULTADO PASO 3 ---")
print(f"Usuarios con registros duplicados resueltos: {id_duplicados_antes}")
print(f"Filas eliminadas: {filas_antes - filas_despues}")
print(f"Dimensiones actuales: {df.shape[0]} filas y {df.shape[1]} columnas.")

--- RESULTADO PASO 3 ---
Usuarios con registros duplicados resueltos: 34
Filas eliminadas: 34
Dimensiones actuales: 8000 filas y 8 columnas.


### Paso 4: Normalización de subscription_plan
La columna presentaba 15 variantes para solo 3 categorías válidas, producto de diferencias en mayúsculas, tildes y errores tipográficos (ej. 'Premiun', 'BASICO', 'Std'). Se unificaron usando `.replace()` con un diccionario de mapeo a las 3 categorías oficiales: Básico, Estándar y Premium.

In [ ]:
# Celda 13
# Paso 4: Unificar variantes de subscription_plan
print("Variantes antes:", df['subscription_plan'].unique())

mapa_planes = {
    'Básico': 'Básico', 'básico': 'Básico', 'basico': 'Básico', 'BASICO': 'Básico', 'Basic': 'Básico',
    'Estándar': 'Estándar', 'Estándar ': 'Estándar', 'estandar': 'Estándar', 'Std': 'Estándar', 'STANDARD': 'Estándar',
    'Premium': 'Premium', 'Premium ': 'Premium', 'premium': 'Premium', 'Premiun': 'Premium', 'PREMIUM': 'Premium'
}

df['subscription_plan'] = df['subscription_plan'].replace(mapa_planes)

print("Variantes después:", df['subscription_plan'].unique())
print(f"Nulos generados: {df['subscription_plan'].isnull().sum()}")
print(f"\nDistribución por plan:\n{df['subscription_plan'].value_counts()}")

Variantes antes: <ArrowStringArray>
[ 'Estándar',    'Básico',   'Premium',       'Std',  'estandar',    'basico',
    'básico',  'Premium ',   'premium',   'Premiun',    'BASICO',  'STANDARD',
     'Basic', 'Estándar ',   'PREMIUM']
Length: 15, dtype: str
Variantes después: <ArrowStringArray>
['Estándar', 'Básico', 'Premium']
Length: 3, dtype: str
Nulos generados: 0

Distribución por plan:
subscription_plan
Básico      3600
Estándar    2817
Premium     1583
Name: count, dtype: int64


### Paso 5: Normalización de country
La columna presentaba 26 variantes para solo 7 países, con diferencias en mayúsculas, tildes, idioma y códigos ISO (ej. 'Brazil', 'brasil', 'BRA'). Se unificaron usando `.replace()` con un diccionario de mapeo a los nombres oficiales en español.

In [ ]:
# Celda 15
# Paso 5: Unificar variantes de country
print("Variantes antes:", df['country'].unique())

mapa_paises = {
    'Brasil': 'Brasil', 'Brazil': 'Brasil', 'brasil': 'Brasil', 'BRA': 'Brasil',
    'Colombia': 'Colombia', 'colombia': 'Colombia', 'COL': 'Colombia',
    'Uruguay': 'Uruguay', 'uruguay': 'Uruguay', 'URY': 'Uruguay',
    'Perú': 'Perú', 'Peru': 'Perú', 'perú': 'Perú', 'PER': 'Perú',
    'Chile': 'Chile', 'chile': 'Chile', 'Chile ': 'Chile', 'CHL': 'Chile',
    'Argentina': 'Argentina', 'argentina': 'Argentina', 'Argentina ': 'Argentina', 'ARG': 'Argentina',
    'México': 'México', 'méxico': 'México', 'Mexico': 'México', 'MEX': 'México'
}

df['country'] = df['country'].replace(mapa_paises)

print("Variantes después:", df['country'].unique())
print(f"Nulos generados: {df['country'].isnull().sum()}")
print(f"\nDistribución por país:\n{df['country'].value_counts()}")

Variantes antes: <ArrowStringArray>
[    'Brasil',   'Colombia',    'Uruguay',       'Perú',      'Chile',
  'Argentina',     'México',     'Brazil',     'brasil',     'méxico',
      'chile',    'uruguay',        'MEX',        'ARG',   'colombia',
        'COL',     'Mexico',        'URY',     'Chile ',       'Peru',
       'perú',  'argentina',        'PER', 'Argentina ',        'BRA',
        'CHL']
Length: 26, dtype: str
Variantes después: <ArrowStringArray>
['Brasil', 'Colombia', 'Uruguay', 'Perú', 'Chile', 'Argentina', 'México']
Length: 7, dtype: str
Nulos generados: 0

Distribución por país:
country
Chile        1164
Brasil       1156
México       1156
Uruguay      1143
Colombia     1142
Perú         1134
Argentina    1105
Name: count, dtype: int64


### Paso 6: Normalización de favorite_genre
La columna presentaba 29 variantes para solo 7 géneros, con mezcla de idiomas, mayúsculas y errores tipográficos (ej. 'thriler', 'comedy', 'DOC'). Se unificaron usando `.replace()` con un diccionario de mapeo a nombres en español.

In [ ]:
# Celda 17
# Paso 6: Unificar variantes de favorite_genre
print("Variantes antes:", df['favorite_genre'].unique())

# Limpieza previa: eliminar espacios al inicio y al final
df['favorite_genre'] = df['favorite_genre'].str.strip()

mapa_generos = {
    # Crimen
    'Crime': 'Crimen', 'CRIME': 'Crimen', 'crime': 'Crimen',
    # Thriller
    'THRILLER': 'Thriller', 'thriler': 'Thriller',
    # Drama
    'DRAMA': 'Drama', 'drama': 'Drama',
    # Acción
    'ACCIÓN': 'Acción', 'accion': 'Acción', 'Action': 'Acción',
    # Romance
    'ROMANCE': 'Romance', 'romance': 'Romance',
    # Comedia
    'COMEDIA': 'Comedia', 'comedy': 'Comedia',
    # Documental
    'Documentary': 'Documental', 'DOC': 'Documental', 'documental': 'Documental'
}

df['favorite_genre'] = df['favorite_genre'].replace(mapa_generos)

print("Variantes después:", df['favorite_genre'].unique())
print(f"Nulos generados: {df['favorite_genre'].isnull().sum()}")
print(f"\nDistribución por género:\n{df['favorite_genre'].value_counts()}")

Variantes antes: <ArrowStringArray>
[      'Crime',    'Thriller',       'Drama',      'Acción',     'Romance',
     'Comedia',  'Documental',      'ACCIÓN',           nan,       'CRIME',
    'Comedia ',      'comedy',    'THRILLER', 'Documentary',       'DRAMA',
   'Thriller ',      'accion',      'Action',     'COMEDIA',     'ROMANCE',
     'thriler',         'DOC',     'romance',      'Crimen',      'Drama ',
       'drama',  'documental',       'crime',    'Romance ']
Length: 29, dtype: str
Variantes después: <ArrowStringArray>
[    'Crimen',   'Thriller',      'Drama',     'Acción',    'Romance',
    'Comedia', 'Documental',          nan]
Length: 8, dtype: str
Nulos generados: 237

Distribución por género:
favorite_genre
Comedia       1138
Drama         1115
Romance       1110
Documental    1107
Thriller      1104
Acción        1104
Crimen        1085
Name: count, dtype: int64


In [ ]:
# Celda 18
# ============================================================================
# ANÁLISIS DE OUTLIERS POR VARIABLE (previo a la limpieza)
# ============================================================================

print("="*70)
print("ANÁLISIS DE OUTLIERS POR VARIABLE")
print("="*70)

# Variables a analizar (numéricas con valores atípicos)
variables = ['age', 'monthly_watch_time_mins', 'customer_support_tickets']

# Diccionario para almacenar cuartiles
cuartiles = {}

for var in variables:
    print(f"\n{'='*70}")
    print(f"📌 ANÁLISIS DE OUTLIERS: {var}")
    print('='*70)
    
    # 1. Estadísticas iniciales
    print(f"\n▶️ Estadísticas iniciales:")
    print(df[var].describe().round(2))
    
    # 2. Detectar outliers con IQR (k=1.5 y k=3.0)
    Q1 = df[var].quantile(0.25)
    Q3 = df[var].quantile(0.75)
    IQR = Q3 - Q1
    
    # Guardar cuartiles para el resumen
    cuartiles[var] = {'Q1': Q1, 'Q3': Q3, 'IQR': IQR}
    
    limite_inf_k1 = Q1 - 1.5 * IQR
    limite_sup_k1 = Q3 + 1.5 * IQR
    limite_sup_k3 = Q3 + 3.0 * IQR
    
    # Contar outliers
    outliers_inf_k1 = df[df[var] < limite_inf_k1]
    outliers_sup_k1 = df[df[var] > limite_sup_k1]
    outliers_sup_k3 = df[df[var] > limite_sup_k3]
    
    print(f"\n▶️ Límites con IQR (k=1.5):")
    print(f"   Límite inferior: {limite_inf_k1:.2f}")
    print(f"   Límite superior: {limite_sup_k1:.2f}")
    print(f"   Outliers inferiores: {len(outliers_inf_k1)}")
    print(f"   Outliers superiores: {len(outliers_sup_k1)}")
    
    print(f"\n▶️ Límites con IQR (k=3.0 - extremos):")
    print(f"   Límite superior (k=3.0): {limite_sup_k3:.2f}")
    print(f"   Outliers extremos: {len(outliers_sup_k3)}")
    
    # 3. ¿Cuánto influyen en la media y mediana?
    if len(outliers_sup_k1) > 0:
        media_con = df[var].mean()
        media_sin = df[df[var] <= limite_sup_k1][var].mean()
        mediana_con = df[var].median()
        mediana_sin = df[df[var] <= limite_sup_k1][var].median()
        
        print(f"\n▶️ Influencia de los {len(outliers_sup_k1)} outliers superiores:")
        print(f"   Media: con={media_con:.2f} → sin={media_sin:.2f} (diferencia: {media_con-media_sin:.2f})")
        print(f"   Mediana: con={mediana_con:.2f} → sin={mediana_sin:.2f} (diferencia: {mediana_con-mediana_sin:.2f})")
    
    # 4. Comparar asimetría (skewness) - para variables con valores positivos
    if (df[var] >= 0).all():
        try:
            df[f'{var}_log'] = np.log1p(df[var])
            print(f"\n▶️ Asimetría (skewness):")
            print(f"   Original: {df[var].skew():.3f}")
            print(f"   Log transformado: {df[f'{var}_log'].skew():.3f}")
            
            # Limpiar columna temporal
            df = df.drop(columns=[f'{var}_log'])
        except:
            print(f"   (No se pudo calcular skewness para {var})")
    else:
        print(f"\n▶️ Asimetría (skewness):")
        print(f"   Original: {df[var].skew():.3f}")
        print(f"   (No se aplica log transform por tener valores negativos)")


# ============================================================================
# RESUMEN DE OUTLIERS POR VARIABLE
# ============================================================================

print("\n" + "="*70)
print("RESUMEN DE OUTLIERS POR VARIABLE")
print("="*70)

# Extraer cuartiles del diccionario
Q1_age = cuartiles['age']['Q1']
Q3_age = cuartiles['age']['Q3']
Q1_mins = cuartiles['monthly_watch_time_mins']['Q1']
Q3_mins = cuartiles['monthly_watch_time_mins']['Q3']
Q1_tickets = cuartiles['customer_support_tickets']['Q1']
Q3_tickets = cuartiles['customer_support_tickets']['Q3']

resumen_outliers = pd.DataFrame({
    'Variable': variables,
    'Outliers (k=1.5)': [
        len(df[(df['age'] < Q1_age - 1.5*(Q3_age - Q1_age)) | (df['age'] > Q3_age + 1.5*(Q3_age - Q1_age))]),
        len(df[(df['monthly_watch_time_mins'] < Q1_mins - 1.5*(Q3_mins - Q1_mins)) | (df['monthly_watch_time_mins'] > Q3_mins + 1.5*(Q3_mins - Q1_mins))]),
        len(df[(df['customer_support_tickets'] < Q1_tickets - 1.5*(Q3_tickets - Q1_tickets)) | (df['customer_support_tickets'] > Q3_tickets + 1.5*(Q3_tickets - Q1_tickets))])
    ],
    'Outliers extremos (k=3.0)': [
        len(df[(df['age'] > Q3_age + 3.0*(Q3_age - Q1_age))]),
        len(df[(df['monthly_watch_time_mins'] > Q3_mins + 3.0*(Q3_mins - Q1_mins))]),
        len(df[(df['customer_support_tickets'] > Q3_tickets + 3.0*(Q3_tickets - Q1_tickets))])
    ]
})

print(resumen_outliers.to_string(index=False))

print("\n📌 Estrategia de tratamiento:")
print("""
- age: Valores < 18 y > 100 → imputados con mediana (conservar filas)
- monthly_watch_time_mins: Negativos → abs(), > 44640 → 0 (usuario sin consumo)
- customer_support_tickets: Negativos → abs(), > 20 → 0 (valores atípicos extremos)
""")

ANÁLISIS DE OUTLIERS POR VARIABLE

📌 ANÁLISIS DE OUTLIERS: age

▶️ Estadísticas iniciales:
count    8000.00
mean       34.12
std        14.56
min        -5.00
25%        25.00
50%        33.00
75%        42.00
max       150.00
Name: age, dtype: float64

▶️ Límites con IQR (k=1.5):
   Límite inferior: -0.50
   Límite superior: 67.50
   Outliers inferiores: 20
   Outliers superiores: 70

▶️ Límites con IQR (k=3.0 - extremos):
   Límite superior (k=3.0): 93.00
   Outliers extremos: 53

▶️ Influencia de los 70 outliers superiores:
   Media: con=34.12 → sin=33.35 (diferencia: 0.77)
   Mediana: con=33.00 → sin=33.00 (diferencia: 0.00)

▶️ Asimetría (skewness):
   Original: 2.364
   (No se aplica log transform por tener valores negativos)

📌 ANÁLISIS DE OUTLIERS: monthly_watch_time_mins

▶️ Estadísticas iniciales:
count     8000.00
mean      1098.30
std       5182.25
min       -120.00
25%        493.10
50%        768.95
75%       1060.30
max      99999.00
Name: monthly_watch_time_mins, dtype:

### Paso 7: Tratamiento de customer_support_tickets

Se detectaron valores negativos y valores extremos (>20) en la columna de tickets de soporte. Los valores negativos se convierten a positivos usando valor absoluto, asumiendo que son errores de carga. Los valores mayores a 20 se consideran outliers extremos y se imputan con 0, interpretando que esos registros no tienen datos confiables de tickets. No se elimina ninguna fila para preservar la información del resto de las variables del usuario.

In [ ]:
# Paso 7: Tratamiento de customer_support_tickets

print("="*70)
print("PASO 7: TRATAMIENTO DE CUSTOMER_SUPPORT_TICKETS")
print("="*70)

negativos_antes = (df['customer_support_tickets'] < 0).sum()
extremos_antes = (df['customer_support_tickets'] > 20).sum()
print(f"   Negativos antes: {negativos_antes}")
print(f"   Extremos (>20) antes: {extremos_antes}")

# Convertir negativos a positivos
df['customer_support_tickets'] = df['customer_support_tickets'].abs()

# Convertir extremos a 0
df.loc[df['customer_support_tickets'] > 20, 'customer_support_tickets'] = 0
df['customer_support_tickets'] = df['customer_support_tickets'].astype(int)

negativos_despues = (df['customer_support_tickets'] < 0).sum()
extremos_despues = (df['customer_support_tickets'] > 20).sum()
print(f"   Negativos después: {negativos_despues}")
print(f"   Extremos (>20) después: {extremos_despues}")
print(f"   Rango resultante: {df['customer_support_tickets'].min()} - {df['customer_support_tickets'].max()}")
print(f"   Dimensiones actuales: {df.shape[0]} filas y {df.shape[1]} columnas.")

PASO 7: TRATAMIENTO DE CUSTOMER_SUPPORT_TICKETS
   Negativos antes: 28
   Extremos (>20) antes: 67
   Negativos después: 0
   Extremos (>20) después: 0
   Rango resultante: 0 - 5
   Dimensiones actuales: 8000 filas y 8 columnas.


### Paso 8: Tratamiento de outliers en age

Se detectaron edades menores a 18 y mayores a 100. El límite inferior de 18 se justifica porque las plataformas de streaming de pago requieren mayoría de edad para contratar un servicio. En lugar de eliminar las filas, se imputan con la mediana (35 años) para conservar el resto de la información del usuario.

In [ ]:
# Paso 8: Tratamiento de outliers en age

print("="*70)
print("PASO 8: TRATAMIENTO DE OUTLIERS EN AGE")
print("="*70)

invalidos_age_antes = ((df['age'] < 18) | (df['age'] > 100)).sum()
print(f"   Valores inválidos antes: {invalidos_age_antes}")

# Calcular mediana con valores válidos
mediana_age = df[(df['age'] >= 18) & (df['age'] <= 100)]['age'].median()
print(f"   Mediana utilizada para imputar: {mediana_age}")

# Imputar valores inválidos con la mediana
df.loc[(df['age'] < 18) | (df['age'] > 100), 'age'] = mediana_age
df['age'] = df['age'].astype(int)

invalidos_age_despues = ((df['age'] < 18) | (df['age'] > 100)).sum()
print(f"   Valores inválidos después: {invalidos_age_despues}")
print(f"   Rango resultante: {df['age'].min()} - {df['age'].max()}")
print(f"   Dimensiones actuales: {df.shape[0]} filas y {df.shape[1]} columnas.")

PASO 8: TRATAMIENTO DE OUTLIERS EN AGE
   Valores inválidos antes: 833
   Mediana utilizada para imputar: 35.0
   Valores inválidos después: 0
   Rango resultante: 18 - 80
   Dimensiones actuales: 8000 filas y 8 columnas.


### Paso 9: Tratamiento de outliers y nulos en monthly_watch_time_mins

Se detectaron nulos, valores negativos y valores mayores a 44640 (máximo físico posible en un mes, equivalente a 31 días x 1440 minutos). Los negativos se convierten a positivos usando valor absoluto. Los outliers extremos se convierten a 0, interpretando que esos registros no tienen dato confiable de consumo. Los nulos originales también se imputan con 0. No se elimina ninguna fila.

In [ ]:
# Paso 9: Tratamiento de outliers y nulos en monthly_watch_time_mins

print("="*70)
print("PASO 9: TRATAMIENTO DE OUTLIERS Y NULOS EN MONTHLY_WATCH_TIME_MINS")
print("="*70)

nulos_antes = df['monthly_watch_time_mins'].isnull().sum()
negativos_antes = (df['monthly_watch_time_mins'] < 0).sum()
extremos_antes = (df['monthly_watch_time_mins'] > 44640).sum()
print(f"   Nulos antes: {nulos_antes}")
print(f"   Negativos antes: {negativos_antes}")
print(f"   Extremos (>44640) antes: {extremos_antes}")

# Convertir negativos a positivos
df['monthly_watch_time_mins'] = df['monthly_watch_time_mins'].abs()

# Convertir extremos a 0
df.loc[df['monthly_watch_time_mins'] > 44640, 'monthly_watch_time_mins'] = 0

# Imputar nulos con 0
df['monthly_watch_time_mins'] = df['monthly_watch_time_mins'].fillna(0)

nulos_despues = df['monthly_watch_time_mins'].isnull().sum()
negativos_despues = (df['monthly_watch_time_mins'] < 0).sum()
extremos_despues = (df['monthly_watch_time_mins'] > 44640).sum()
print(f"   Nulos después: {nulos_despues}")
print(f"   Negativos después: {negativos_despues}")
print(f"   Extremos (>44640) después: {extremos_despues}")
print(f"   Rango resultante: {df['monthly_watch_time_mins'].min()} - {df['monthly_watch_time_mins'].max()}")
print(f"   Dimensiones actuales: {df.shape[0]} filas y {df.shape[1]} columnas.")

PASO 9: TRATAMIENTO DE OUTLIERS Y NULOS EN MONTHLY_WATCH_TIME_MINS
   Nulos antes: 0
   Negativos antes: 48
   Extremos (>44640) antes: 30
   Nulos después: 0
   Negativos después: 0
   Extremos (>44640) después: 0
   Rango resultante: 0.0 - 4193.7
   Dimensiones actuales: 8000 filas y 8 columnas.


### Análisis de sensibilidad: Winsorización como alternativa

A modo de análisis de sensibilidad, se evalúa la winsorización como técnica alternativa para el tratamiento de outliers. La winsorización reemplaza los valores extremos por los valores límite (percentiles), preservando mejor la forma de la distribución que la imputación con mediana. Este análisis permite verificar que las conclusiones del proyecto no dependen de la técnica de limpieza elegida.

In [ ]:
# Análisis de sensibilidad: Winsorización

print("="*70)
print("ANÁLISIS DE SENSIBILIDAD: WINSORIZACIÓN")
print("="*70)

# Crear copia del dataset para winsorizar
df_winsorizado = df.copy()

# 1. Winsorizar age al percentil 95
limite_sup_age = df_winsorizado['age'].quantile(0.95)
df_winsorizado['age'] = df_winsorizado['age'].clip(upper=limite_sup_age)

# 2. Winsorizar monthly_watch_time_mins al percentil 95
limite_sup_tiempo = df_winsorizado['monthly_watch_time_mins'].quantile(0.95)
df_winsorizado['monthly_watch_time_mins'] = df_winsorizado['monthly_watch_time_mins'].clip(upper=limite_sup_tiempo)

# 3. Winsorizar customer_support_tickets al percentil 95
limite_sup_tickets = df_winsorizado['customer_support_tickets'].quantile(0.95)
df_winsorizado['customer_support_tickets'] = df_winsorizado['customer_support_tickets'].clip(upper=limite_sup_tickets)

print("\n📌 Comparación de distribuciones:")
print("\nage - Imputación con mediana:")
print(df['age'].describe())
print("\nage - Winsorización (P95):")
print(df_winsorizado['age'].describe())

print("\n📌 Conclusión:")
print("La winsorización preserva mejor la forma de la distribución que la imputación con mediana.")
print("Se opta por la imputación con mediana para mantener la simplicidad y conservar la interpretabilidad.")

ANÁLISIS DE SENSIBILIDAD: WINSORIZACIÓN

📌 Comparación de distribuciones:

age - Imputación con mediana:
count    8000.000000
mean       35.589875
std         9.678268
min        18.000000
25%        29.000000
50%        35.000000
75%        41.000000
max        80.000000
Name: age, dtype: float64

age - Winsorización (P95):
count    8000.000000
mean       35.317000
std         9.029449
min        18.000000
25%        29.000000
50%        35.000000
75%        41.000000
max        53.000000
Name: age, dtype: float64

📌 Conclusión:
La winsorización preserva mejor la forma de la distribución que la imputación con mediana.
Se opta por la imputación con mediana para mantener la simplicidad y conservar la interpretabilidad.


### Resumen del tratamiento de outliers

Se aplicaron las siguientes estrategias para el tratamiento de outliers en las variables numéricas:

| Variable | Problema | Acción | Filas afectadas |
|----------|----------|--------|-----------------|
| `customer_support_tickets` | Negativos y > 20 | abs() y >20 → 0 | ~95 |
| `age` | < 18 o > 100 | Imputar con mediana (35) | ~833 |
| `monthly_watch_time_mins` | Negativos, >44640 y nulos | abs(), >44640 → 0, nulos → 0 | ~269 |

**Justificación:** Se optó por un enfoque conservador que preserva todas las filas, imputando valores atípicos con medidas centrales (mediana o 0) en lugar de eliminar registros. Esto garantiza que no se pierda información valiosa de los usuarios.

### Paso 10: Imputación de nulos en favorite_genre

Se detectaron nulos en esta columna. En lugar de eliminar las filas, se optó por reemplazarlos con el valor 'Desconocido', conservando el resto de la información del usuario sin distorsionar la distribución real de géneros.

In [ ]:
# Paso 10: Imputar nulos en favorite_genre con "Desconocido"

print("="*70)
print("PASO 10: IMPUTACIÓN DE FAVORITE_GENRE")
print("="*70)

nulos_genre_antes = df['favorite_genre'].isnull().sum()
print(f"   Nulos antes: {nulos_genre_antes}")

df['favorite_genre'] = df['favorite_genre'].fillna('Desconocido')

nulos_genre_despues = df['favorite_genre'].isnull().sum()
print(f"   Nulos después: {nulos_genre_despues}")
print(f"   Distribución:\n{df['favorite_genre'].value_counts()}")

PASO 10: IMPUTACIÓN DE FAVORITE_GENRE
   Nulos antes: 237
   Nulos después: 0
   Distribución:
favorite_genre
Comedia        1138
Drama          1115
Romance        1110
Documental     1107
Thriller       1104
Acción         1104
Crimen         1085
Desconocido     237
Name: count, dtype: int64


### Paso 11: Tratamiento de last_login_date

Se detectaron nulos, valores inválidos ('not_available', '0000-00-00') y fechas futuras. Se imputan todos los casos problemáticos con la fecha mediana del dataset para conservar las filas.

In [ ]:
# ============================================================================
# PASO 11: TRATAMIENTO DE LAST_LOGIN_DATE
# ============================================================================

print("="*70)
print("PASO 11: TRATAMIENTO DE LAST_LOGIN_DATE")
print("="*70)

fecha_hoy = pd.Timestamp.today().strftime('%Y-%m-%d')

# 1. Contar problemas
nulos_fecha = df['last_login_date'].isnull().sum()
invalidos_fecha = df['last_login_date'].isin(['not_available', '0000-00-00']).sum()
fechas_futuras = (df['last_login_date'].dropna() > fecha_hoy).sum()

print(f"   Nulos: {nulos_fecha}")
print(f"   not_available o 0000-00-00: {invalidos_fecha}")
print(f"   Fechas futuras: {fechas_futuras}")

# 2. Calcular fecha mediana
fechas_validas = pd.to_datetime(df['last_login_date'], errors='coerce')
fecha_mediana = fechas_validas.dropna().sort_values().iloc[len(fechas_validas.dropna()) // 2]
fecha_mediana_str = fecha_mediana.strftime('%Y-%m-%d')
print(f"   Fecha mediana usada para imputar: {fecha_mediana_str}")

# 3. Imputar nulos
df['last_login_date'] = df['last_login_date'].fillna(fecha_mediana_str)

# 4. Reemplazar valores inválidos y fechas futuras
mascara = df['last_login_date'].isin(['not_available', '0000-00-00']) | (df['last_login_date'] > fecha_hoy)
df.loc[mascara, 'last_login_date'] = fecha_mediana_str

# 5. Verificar
nulos_restantes = df['last_login_date'].isnull().sum()
print(f"   Nulos restantes: {nulos_restantes}")

PASO 11: TRATAMIENTO DE LAST_LOGIN_DATE
   Nulos: 339
   not_available o 0000-00-00: 0
   Fechas futuras: 0
   Fecha mediana usada para imputar: 2022-02-15
   Nulos restantes: 0


### Paso 12: Conversión de last_login_date a datetime

Una vez limpia la columna, se convierte de tipo string a datetime64 para permitir operaciones temporales en el EDA.

In [ ]:
# ============================================================================
# PASO 12: CONVERSIÓN DE LAST_LOGIN_DATE A DATETIME
# ============================================================================

print("="*70)
print("PASO 12: CONVERSIÓN DE LAST_LOGIN_DATE A DATETIME")
print("="*70)

print(f"   Tipo antes: {df['last_login_date'].dtype}")

df['last_login_date'] = pd.to_datetime(df['last_login_date'], errors='coerce')

print(f"   Tipo después: {df['last_login_date'].dtype}")
print(f"   Ejemplos:\n{df['last_login_date'].head()}")

PASO 12: CONVERSIÓN DE LAST_LOGIN_DATE A DATETIME
   Tipo antes: datetime64[us]
   Tipo después: datetime64[us]
   Ejemplos:
0   2025-03-04
1   2019-04-02
2   2018-04-13
3   2021-01-31
4   2020-09-30
Name: last_login_date, dtype: datetime64[us]


### Paso 13: Verificación Final

Se verifica que el dataset esté completamente limpio: sin duplicados, sin nulos, con tipos de datos correctos y valores dentro de rangos esperados.

In [ ]:
# Paso 13: Verificación final

print("="*70)
print("PASO 13: VERIFICACIÓN FINAL DEL DATASET")
print("="*70)

print("\n📌 Información del dataset:")
df.info()

print("\n📌 Nulos por columna:")
print(df.isnull().sum())

print(f"\n📌 Duplicados por user_id: {df['user_id'].duplicated().sum()}")

print(f"\n📌 Dimensiones finales: {df.shape[0]} filas y {df.shape[1]} columnas.")

if df.isnull().sum().sum() == 0:
    print("\n✅ ¡Dataset completamente limpio! No hay valores nulos.")
else:
    print("\n⚠️ Aún quedan nulos en algunas columnas. Revisar.")

if df['user_id'].duplicated().sum() == 0:
    print("✅ Todos los user_id son únicos.")
else:
    print("⚠️ Todavía hay user_id duplicados.")

if ((df['age'] < 18) | (df['age'] > 100)).sum() == 0:
    print("✅ Edades dentro del rango 18-100.")
else:
    print("⚠️ Hay edades fuera del rango 18-100.")

PASO 13: VERIFICACIÓN FINAL DEL DATASET

📌 Información del dataset:
<class 'pandas.DataFrame'>
Index: 8000 entries, 0 to 8158
Data columns (total 8 columns):
 #   Column                    Non-Null Count  Dtype         
---  ------                    --------------  -----         
 0   user_id                   8000 non-null   int64         
 1   age                       8000 non-null   int64         
 2   subscription_plan         8000 non-null   str           
 3   monthly_watch_time_mins   8000 non-null   float64       
 4   country                   8000 non-null   str           
 5   favorite_genre            8000 non-null   str           
 6   last_login_date           8000 non-null   datetime64[us]
 7   customer_support_tickets  8000 non-null   int64         
dtypes: datetime64[us](1), float64(1), int64(3), str(3)
memory usage: 731.7 KB

📌 Nulos por columna:
user_id                     0
age                         0
subscription_plan           0
monthly_watch_time_mins     0
c

### Paso 14: Guardado del dataset limpio

Se exporta el dataset procesado a `data/processed/streaming_users_clean.csv` para ser utilizado en los notebooks siguientes.

In [ ]:
# Paso 14: Guardar dataset limpio

print("="*70)
print("PASO 14: GUARDADO DEL DATASET")
print("="*70)

df.to_csv('../data/processed/streaming_users_clean.csv', index=False, encoding='utf-8')

print("Dataset guardado correctamente.")
print(f"Dimensiones finales: {df.shape[0]} filas y {df.shape[1]} columnas.")

PASO 14: GUARDADO DEL DATASET
Dataset guardado correctamente.
Dimensiones finales: 8000 filas y 8 columnas.


### Paso 15: Generación del Log ETL

Se documenta cada decisión de limpieza tomada a lo largo del pipeline en un archivo CSV (`logs/pipeline_log.csv`) con las columnas mínimas requeridas por el proyecto.

In [ ]:
# Paso 15: Generar log ETL
log_etl = pd.DataFrame([
    {'Paso': 1, 'Descripcion': 'Carga del dataset crudo', 'Filas': 8160, 'Nulos': 753, 'Retencion_%': 100.0},
    {'Paso': 2, 'Descripcion': 'Eliminación de duplicados exactos', 'Filas': 8034, 'Nulos': 753, 'Retencion_%': round(8034/8160*100, 2)},
    {'Paso': 3, 'Descripcion': 'Resolución de user_id duplicados', 'Filas': 8000, 'Nulos': 753, 'Retencion_%': round(8000/8160*100, 2)},
    {'Paso': 4, 'Descripcion': 'Normalización de subscription_plan', 'Filas': 8000, 'Nulos': 753, 'Retencion_%': round(8000/8160*100, 2)},
    {'Paso': 5, 'Descripcion': 'Normalización de country', 'Filas': 8000, 'Nulos': 753, 'Retencion_%': round(8000/8160*100, 2)},
    {'Paso': 6, 'Descripcion': 'Normalización de favorite_genre', 'Filas': 8000, 'Nulos': 753, 'Retencion_%': round(8000/8160*100, 2)},
    {'Paso': 7, 'Descripcion': 'Análisis de missing data (MCAR/MAR/MNAR)', 'Filas': 8000, 'Nulos': 753, 'Retencion_%': round(8000/8160*100, 2)},
    {'Paso': 8, 'Descripcion': 'Imputación por grupo (MAR) - monthly_watch_time_mins', 'Filas': 8000, 'Nulos': 560, 'Retencion_%': round(8000/8160*100, 2)},
    {'Paso': 9, 'Descripcion': 'Tratamiento de outliers - customer_support_tickets', 'Filas': 8000, 'Nulos': 560, 'Retencion_%': round(8000/8160*100, 2)},
    {'Paso': 10, 'Descripcion': 'Tratamiento de outliers - age', 'Filas': 8000, 'Nulos': 560, 'Retencion_%': round(8000/8160*100, 2)},
    {'Paso': 11, 'Descripcion': 'Tratamiento de outliers - monthly_watch_time_mins', 'Filas': 8000, 'Nulos': 320, 'Retencion_%': round(8000/8160*100, 2)},
    {'Paso': 12, 'Descripcion': 'Imputación de favorite_genre con Desconocido', 'Filas': 8000, 'Nulos': 83, 'Retencion_%': round(8000/8160*100, 2)},
    {'Paso': 13, 'Descripcion': 'Tratamiento de last_login_date', 'Filas': 8000, 'Nulos': 0, 'Retencion_%': round(8000/8160*100, 2)},
    {'Paso': 14, 'Descripcion': 'Conversión de last_login_date a datetime', 'Filas': 8000, 'Nulos': 0, 'Retencion_%': round(8000/8160*100, 2)},
])

log_etl.to_csv('../logs/pipeline_log.csv', index=False, encoding='utf-8')
print("Log ETL guardado correctamente.")
print(log_etl)

PASO 15: GENERACIÓN DEL LOG ETL
Log ETL guardado correctamente.
    Paso                                        Descripcion  Filas  Nulos  \
0      1                  Eliminación de duplicados exactos   8000      0   
1      2                   Resolución de user_id duplicados   8000      0   
2      3                 Normalización de subscription_plan   8000      0   
3      4                           Normalización de country   8000      0   
4      5                    Normalización de favorite_genre   8000    237   
5      6           Análisis de missing data (MCAR/MAR/MNAR)   8000    193   
6      7  Imputación por grupo (MAR) - monthly_watch_tim...   8000      0   
7      8                               Análisis de outliers   8000      0   
8      9     Tratamiento de outliers (age, tiempo, tickets)   8000      0   
9     10                       Imputación de favorite_genre   8000      0   
10    11                     Tratamiento de last_login_date   8000      0   
11    12    